WIP - Adds build process (Gulp.js-based)

snostorm opened the pull request:
For some (easier) transparency in to the progress of this branch I've opened a PR. As of today I assume it is not in a merge-ready state but I did want to get easier access for us to review the ongoing diff and moving towards a merge sooner-than-later.

I don't think we need to get everything perfect early on, just solid/simple enough with a reasonable enough pattern we can build some additional content off of.

therebelrobot committed:
Added package.json, gulpfile and .gitignore

therebelrobot committed:
Added gulp-markdown gulp-minify-css gulp-htmlmin gulp-less and browserify for gulpfile inclusion

therebelrobot committed:
Added gulp tasks and config files for tasks, added content and source folders

snostorm committed:
Merge branch 'master' into build/gulpjs

Fishrock123 committed:
use postcss for css autoprefixing and minification

Fishrock123 committed:
build: use gulp-stylus

therebelrobot committed:
Clean task build out

therebelrobot committed:
Merge branch 'master' into build/gulpjs

therebelrobot committed:
Stylus Compiling now working

therebelrobot committed:
Build/minifyCSS working now

Fishrock123 committed:
translate styles.css to stylus

therebelrobot committed:
Templating / i18n working

therebelrobot committed:
Merge pull request #100 from iojs/build/stylus

translate styles.css to stylus

therebelrobot committed:
[minor] removed unneeded object holders, added in noDuplicate param for if Substack merges my commit over there.

therebelrobot committed:
Merge branch 'build/gulpjs' of github.com:iojs/website into build/gulpjs

therebelrobot committed:
Merge branch 'master' into build/gulpjs

snostorm committed:
Merge branch 'master' into build/gulpjs

therebelrobot committed:
Added some templates for current content

snostorm commented:
What are the current blockers? Maybe we can edit the PR description at the top to include a checklist.

therebelrobot committed:
Merge branch 'build/gulpjs' of github.com:iojs/website into build/gulpjs

therebelrobot committed:
Merge branch 'master' into build/gulpjs

snostorm commented:
Need any help today @therebelrobot ? Whether with review, testing or breaking off some sub tasks?

therebelrobot committed:
Added template for main page, also added sample for spanish to test i18n

therebelrobot committed:
Removed interim gulp testing

therebelrobot committed:
Added a page identifier to the main template for css triggering

Also moved the index css to the bottom of the stylus page, made it page
specific

therebelrobot committed:
split out individual page css to ensure no cross-page issues

specifically with the generic nature of the content section

mikeal commented:
Great work, and thanks for rushing this cause HOLY SHIT THERE ARE SO MANY PEOPLE DOING LOCALIZATIONS!

therebelrobot committed:
Added basic content for es6

therebelrobot committed:
Removed error in markdown translation of es6 page

therebelrobot committed:
Added mobile.styl for @media query styling

therebelrobot committed:
Added english faq to spanish for translation

therebelrobot committed:
Added page title to template.json

therebelrobot committed:
Updated styling for es6, added es6 placeholder in content/es

therebelrobot committed:
Readying for merge

- changed path in gulpfile to point to public instead of public_test
- updated a small issue with firefox css rendering
- changed logo link to point to index file directly

therebelrobot committed:
Updated build task to disable minifying for the time being

snostorm commented:
Reviewing and properly responding in the next 20-40m @therebelrobot, thanks!

therebelrobot committed:
Added placeholder folders for new languages

therebelrobot committed:
And the build files for the languages

snostorm commented:
@therebelrobot noted. Pulling down now. Will also pop in IRC.

snostorm commented:
FYI I am adding to the README a little as I go

Fishrock123 commented:
Looking..

Fishrock123 commented on package.json:
```diff
@@ -0,0 +1,65 @@
+    {
+      "name": "Zeke Sikelianos",
+      "github": "zeke"
+    }
```
these are collapse-able to `Name <email> (website)`

Fishrock123 commented on source/styles/index.styl:
```diff
@@ -0,0 +1,70 @@
+[markdown-page="index"]
```
`.page-<page>` is a much faster selector. no need to use attr selectors unless javascript is modifying it. :)

Fishrock123 commented on source/templates/main.html:
```diff
@@ -0,0 +1,53 @@
+      <a href="https://github.com/iojs/io.js/issues" i18n-issues-link >GitHub Issues</a><!--
+   --><a href="https://github.com/iojs" i18n-org-link >GitHub Org</a><!--
+   --><a href="https://webchat.freenode.net/?channels=io.js" i18n-irc-link >IRC Chat</a><!--
+   --><a href="http://logs.libuv.org/io.js/latest" i18n-logs-link >Logs</a><!--
```
`i18n-irc-logs-link` might be better, nit though

Fishrock123 commented on source/templates/main.html:
```diff
@@ -0,0 +1,53 @@
+  <meta property="og:image:type" content="image/png">
+  <meta property="og:image:width" content="1369">
+  <meta property="og:image:height" content="1563">
+  <!-- <a href="https://plus.google.com/101986715696875566237" rel="publisher">Google+</a> -->
```
I think this got completely removed

Fishrock123 commented on source/templates/main.html:
```diff
@@ -0,0 +1,53 @@
+      <div class="spacer"></div>
+      <a href="faq.html" i18n-faq-link >FAQ</a>
+      <a href="es6.html" i18n-es6-link >ES6</a>
+      <a href="https://iojs.org/api/" i18n-api-link >API Docs</a>
```
Let's be consistent with the `attribute>` vs `attribute >`

Fishrock123 commented on gulp/tasks/templates.js:
Probably needs cleanup, but LGTM if it works

Fishrock123 commented:
LGTM. Is there any way to squash this at all?

snostorm commented:
We're not embedding a helper server to help review the site locally? Could be an easy change to add to the default gulp (or a special sub-task) to minimize setup time for our contributors.

(Or, I just missed it.)

Fishrock123 commented:
`gulp-connect` is nice.

snostorm commented:
I can do the server and merge it with the live-reload with my README push, which is almost ready? I'd like it to all be bundled

snostorm committed:
Updates README.md for new build process instructions

snostorm commented:
README pushed for review/corrections. I skipped over the quickly evolving i18n efforts and general contributing/WG/team stuff for now.

therebelrobot committed:
cleaning up

therebelrobot committed:
Merge branch 'build/gulpjs' of github.com:iojs/website into build/gulpjs

therebelrobot committed:
Fixed a stupid error

snostorm committed:
Adds local dev server (wip)

snostorm commented:
@therebelrobot sounds great. The server is there, but I am doing one more commit to make it work more magically with browser LR.

snostorm commented:
I have LR 90% ready but keep getting dragged away from the computer. I can PR post-merge it if the last blocker.

What about the no default index.html (and fallback for es6, faq) pages? Merging probably would break the site...?

therebelrobot committed:
Added root redirect files to save compatibility for now

Also added a TRANSLATION.md file to increase awareness of how to
contribute to translation

mikeal commented:
I'm terrible at squashing :( @Fishrock123 is awesome at it though.

Fishrock123 commented:
If i squash this, it will look like me just copying the directory overtop of the current head and commiting it XD

Fishrock123 commented:
I guess I could try that octopus merge ben told me about..

fengmk2 commented:
Should we ignore all auto build out `*.html` files?

Fishrock123 commented:
Will do the squash and merge tomorrow.

snostorm commented:
@fengmk2 we're including generated .html files in public right now as our web server requires them for publishing purposes. Later we'll be ignoring this, yes.

snostorm commented:
As much as we want to rush this in for speed of facilitating our i18n teams, I'd rather hold on this. After further review/testing the main `templates` task I noticed it was running especially slow for me. Since then I've already had some success with speed gains in the refactoring. I've been moving this file slightly closer to a typical gulp pipeline flow to squeeze out some performance and reduce some file reads/repeated Array parsing.

(The slowness was especially noticeable post watch-trigger while having to wait for the files to reload. On my typically zippy machine I could actually see the HTML appearing in files line-by-line.)

Anyway, I don't want to block this long. If i18n groups are excited to translate, let's get them forking off of the content directory in this branch for the next 24h or so then do a big PR + merge party after this goes live.

mikeal commented:
Don't block this on build performance, having something that is kinda slow is better than having nothing :)

Fishrock123 commented:
Squashed cleanly, how will this interact with the nginx config? Do I need to PR a redirect to /en/ there for now?

Fishrock123 commented:
Oh, that landed in https://github.com/iojs/website/commit/2cdc6c4a522591953103e2181b897627ad20cbd0. Fixing up and landing this.

fengmk2 commented:
:100:

Fishrock123 commented:
@snostorm the css tasks run 22 times slower than the template tasks for me. The template build takes half a second, compared to 11 seconds.

(Also I dunno why the css build is quite that slow..)

Fishrock123 commented:
(unless the templates task end isn't actually at the end of the templates build..)
